docs/RFCS/20210215_per_session_cpu_usage.md

```diff
@@ -0,0 +1,393 @@
+
+```go
+func (ex *connExecutor) run(...) {
+    ...
+    ex.taskGroup = runtime.SetTaskGroup()
+}
+
+...
+
+func (ex *connExecutor) serialize() serverpb.Session {
+    ...
+    return serverpb.Session{
+         ...
+        // Experimental
+        SchedTicks: runtime.GetTaskGroupSchedTicks(ex.taskGroup),
+    }
+}
+```
+
+And then, to make it observable for SQL DBAs, in `sql/crdb_internal.go`:
+
+```go
+func populateSessionsTable(...) {
+   ...
+   ... addRow(
+            ...
+            tree.NewDInt(tree.DInt(session.SchedTicks)),
+   )
+   ...
+}
+```
+
+Example usage:
+
+```
+root@:26257/defaultdb> select node_id, client_address, last_active_Query, sched_ticks 
+                       from crdb_internal.node_sessions;
+
+  node_id | client_address | last_active_query | sched_ticks
+----------+----------------+-------------------+--------------
+        1 | [::1]:27580    | SHOW database     |         972
+        1 | [::1]:27577    | SHOW database     |        6675
+(2 rows)
+```
+
+This shows 2 distinct SQL sessions (there are two interactive terminals
+running a SQL shell connected to this node). The `sched_ticks` column
+reports that one session has used 972 ticks and the other has used 6675 ticks.
+
+The value of sched ticks can be confirmed to increase in relation to
+actual CPU usage. We can compare how the ticks are incremented for several operations:
+
+| Operation                                           | Tick increment                              |
+|-----------------------------------------------------|---------------------------------------------|
+| `pg_sleep(0)`                                       | 20                                          |
+| `pg_sleep(1)`                                       | 20 (sleeping time does not incur CPU usage) |
+| `select count(*) from generate_series(1,10000000)`  | 280                                         |
+| `select count(*) from generate_series(1,100000000)` | 2800  (ten times more than previous)        |
+
+As this confirms, idle sessions don't increase their `sched_ticks`, whereas busy sessions increment
+it in proportion to the work actually done.
+
+(NB: the Go changes outlined above do not include the aggregation of
+sched ticks across multiple nodes when queries are distributed. We'd
+need to think further about how to reason about CPU usage on multiple
+nodes, and whether we even want this to be aggregated, or whether we
+want the CPU ticks to be presented side-by-side for separate nodes.)
+
+PR where this change has been prototyped: https://github.com/cockroachdb/cockroach/pull/60589
+
+## Example metrics: heap usage
+
+The basic concept here is to force the Go heap allocator to use
+different heaps for different task groups.
+
+To achieve this we change the heap allocator structures to attach
+a task group to "pages" (blocks) of system memory that are used
+to carve heap allocations from.
+
+When a heap allocation is performed, the RAM usage counter on the task
```

Comment:
consider trying to introduce an indirection here by tying allocations to "allocators" (like in other languates), not directly to task groups. And so a task group, or individual goroutines, could specify what they current allocator is. So for example userspace could write `allocWidget(allocator) *widget` constructors that allocate with a particular allocator.